### PR TITLE
Refactor: Fix memory leak in strings

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Fixes: https://github.com/lfortran/lfortran/issues/2618, https://github.com/lfortran/lfortran/issues/2619, and https://github.com/lfortran/lfortran/issues/2620

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
This fixes some issues, but I still need to handle the case: character as a return type.
The problem is that I cannot free or deallocate the memory in the function scope as the value is being returned. I need to handle this in calling scope, which seems tricky.

**Comment by certik** [comment]
Right. It's the same issue as with allocatable arrays. The caller owns the pointer, so the caller is responsible for freeing it, or passing it out as a function return value, in which case the new caller owns it.

**Comment by Thirumalai-Shaktivel** [comment]
This fixes the memory leak, but I need to fix the tests failure in integration_tests.

**Comment by Thirumalai-Shaktivel** [comment]
I need to review the changes in `modules_48` and fix last test failure `intrinsics_56`

**Comment by certik** [comment]
You have to resolve the conflict.

**Comment by Thirumalai-Shaktivel** [comment]
Should we allow the following assignment?
```fortran
program expr2
implicit none

character(2), allocatable :: x
x = ''

end program
```
Currently, it seg-faults, and adding `allocate(character :: x)` works.

**Comment by Thirumalai-Shaktivel** [comment]
This PR is ready for review!
I kindly request to review each change carefully in the integration_tests, as I made many changes there.
Current output:
```fortran
program expr2
implicit none

character(:), allocatable :: x
character(len=5) :: y
allocate(character :: x)
x = '1243'
y = '256'
y = '7'
print *, '-', x, '-', len(x)
print *, '-', y, '-', len(y)

end program
```
```console
$ lfortran examples/expr2.f90
- 1243 - 4
- 7     - 5
$ gfortran examples/expr2.f90 && ./a.out
 -1243-           4
 -7    -           5
```

**Comment by Thirumalai-Shaktivel** [comment]
> Should we allow the following assignment?

I thought of allowing it for now, updated here: 9ec0d1681ab0be4f417f7592fed5ca49e95cda76

**Comment by certik** [comment]
Regarding https://github.com/lfortran/lfortran/pull/2632#issuecomment-1765611200, that's rarely used, but I think it means a string of length 2, and the entity is allocatable, so you either allocate the whole string (of length 2) or nothing. And the assignment `x = ''` uses the automatic LHS reallocation to allocate it.

Without our `--realloc-lhs` option we only allow LHS reallocation for strings (since strings otherwise become quite unusable in Fortran), but not other arrays, where it's better to be explicit as a user, to know exactly where allocation will happen.

**Comment by certik** [comment]
The CI has to be fixed first, fastGPT failed to build:
```
[ 47%] Building Fortran object CMakeFiles/gpt2.dir/main.f90.o
double free or corruption (out)
make[2]: *** [CMakeFiles/gpt2.dir/build.make:75: CMakeFiles/gpt2.dir/main.f90.o] Aborted (core dumped)
```

**Comment by Thirumalai-Shaktivel** [comment]
Yup, The issue seems in LFortran asr_to_llvm (std::vector), I'm trying to debug it.

File: integration_tests/case_07.f90 [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-    if (test_case('z') /= "z") error stop
-    if (test_case('k') /= "not found") error stop
-    if (test_case('r') /= "not found") error stop
+    if (trim(test_case('a')) /= "a:j") error stop
+    if (trim(test_case('d')) /= "a:j") error stop
+    if (trim(test_case('l')) /= "l:p") error stop
+    if (trim(test_case('p')) /= "l:p") error stop
+    if (trim(test_case('z')) /= "z") error stop
+    if (trim(test_case('k')) /= "not found") error stop
+    if (trim(test_case('r')) /= "not found") error stop
```
Comment (certik) [inline integration_tests/case_07.f90:9]: We should not be trimming it here.

In Fortran, the string comparison operator works for strings of different lengths. So the original test was correct, we need to test this on strings of different lengths. If you need a simpler test, just add a separate test where you trim it, but we still need the original test as well.

File: integration_tests/file_12.f90 [hunk 2]
```diff
@@ -5,6 +5,7 @@ program file_12
     character(:), allocatable :: sentence1
     integer :: u = 11
     integer :: ios
+    allocate(character :: sentence1)
```
Comment (certik) [inline integration_tests/file_12.f90:8]: This is not needed. Also the allocation is in a way incorrect --- it allocates a string of length 0 ? Writing it this way is a corner case, so you can add a separate test for this, but the original test is correct and that's the most common usage, so we need to keep testing it like that.

File: integration_tests/format_08.f90 [hunk 3]
```diff
@@ -2,8 +2,9 @@ program format_08
     implicit none
     character(:), allocatable :: a
     integer :: b(10)
+    allocate(character :: a)
```
Comment (certik) [inline integration_tests/format_08.f90:5]: Same here, let's remove this.

File: integration_tests/intrinsics_36.f90 [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
-    character(len=25) :: space_at_end = adjustl('gfortran   ')
-    character(len=25) :: space_in_between = adjustl('   g for tran   ')
-    character(len=25) :: spaces_with_symbols = adjustl('  # gfor* t $ ran &    ')
+    character(len=1) :: empty = adjustl('')
+    character(len=4) :: all_spaces = adjustl('    ')
+    character(len=8) :: simple = adjustl('gfortran')
+    character(len=11) :: space_at_start = adjustl('   gfortran')
+    character(len=11) :: space_at_end = adjustl('gfortran   ')
+    character(len=16) :: space_in_between = adjustl('   g for tran   ')
+    character(len=23) :: spaces_with_symbols = adjustl('  # gfor* t $ ran &    ')
```
Comment (certik) [inline integration_tests/intrinsics_36.f90:9]: Original test is correct, but the changes are also good to test, so let's make this a separate test.

File: integration_tests/intrinsics_56.f90 [hunk 5]
```diff
@@ -11,9 +11,9 @@ program intrinsics_56
 
 complex, allocatable :: a5(:, :), a6(:, :), b3(:, :)
 
-character(len=5) :: a7(5, 10)
-character(len=5), allocatable :: a8(:, :)
-character(len=5) :: b4(5, 10)
+character(len=2) :: a7(5, 10)
+character(len=2), allocatable :: a8(:, :)
+character(len=2) :: b4(5, 10)
```
Comment (certik) [inline integration_tests/intrinsics_56.f90:16]: Keep the original test, add a separate test if needed.

File: integration_tests/modules_48.f90 [hunk 6]
```diff
@@ -127,12 +130,12 @@ program modules_48
...
 print *, cat_str_alloc, is_fortran_name(cat_str_alloc)
-if( cat_str_alloc /= "runningmodules_48_1:runningmodules_48_2" ) error stop
+if( trim(cat_str_alloc) /= "runningmodules_48_1:runningmodules_48_2" ) error stop
 if( is_fortran_name(cat_str_alloc) ) error stop
 
 cat_str_alloc = string_cat(string_var)
 print *, cat_str_alloc, is_fortran_name(cat_str_alloc)
-if( cat_str_alloc /= "runningmodules_48_1runningmodules_48_2" ) error stop
+if( trim(cat_str_alloc) /= "runningmodules_48_1runningmodules_48_2" ) error stop
 if( .not. is_fortran_name(cat_str_alloc) ) error stop
```
Comment (certik) [inline integration_tests/modules_48.f90:135]: same here, all these changes should be done in a new separate test, and keep the original.

File: integration_tests/modules_49.f90 [hunk 7]
```diff
@@ -39,6 +39,6 @@ program modules_49
 
 character(len=40) :: name = "--gnu-gfortran-11"
 print *, name, to_fortran_name(name)
-if( to_fortran_name(name) /= "__gnu_gfortran_11" ) error stop
+if( trim(to_fortran_name(name)) /= "__gnu_gfortran_11" ) error stop
```
Comment (certik) [inline integration_tests/modules_49.f90:42]: keep the original as well.

**Comment by certik** [comment]
> I kindly request to review each change carefully in the integration_tests, as I made many changes there.

Keep all the original tests intact, they are good tests that we need to support. Add new tests as needed, if you need to test some feature that is not tested yet.

**Comment by Thirumalai-Shaktivel** [comment]
We initialise strings with ` `(space) to each block in the character, so I had to use trim to test the comparison. Consider an example, so Instead of space I need to use some other char to initialise
```fortran
program expr2
implicit none

character(len=5) :: y
y = '7'
print *, '-', y, '-', len(y)

end program
```
```console
$ lfortran examples/expr2.f90
- 7     - 5
```

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 8]
```diff
@@ -923,6 +923,10 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                     if (v) {
                         ASR::ttype_t* struct_t = ASRUtils::TYPE(ASR::make_Struct_t(al, x.base.base.loc, v));
                         new_arg.m_type = struct_t;
+                    } else if (std::string(name_t->m_id) == "character") {
+                        new_arg.m_len_expr = ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                            al, x.m_args[i].loc, 1, ASRUtils::TYPE(ASR::make_Integer_t(
+                            al, x.m_args[i].loc, 4))));
```
Comment (Thirumalai-Shaktivel) [inline src/lfortran/semantics/ast_body_visitor.cpp:929]: This case is used for `character :: x`, which shouldn't be allowed, right?

Comment (certik) [reply src/lfortran/semantics/ast_body_visitor.cpp:929]: Can you post a full example? I am not sure from just this short context.

Comment (Thirumalai-Shaktivel) [reply src/lfortran/semantics/ast_body_visitor.cpp:929]: I'm sorry, here is the minimal example
```fortran
program example
    implicit none
    character(:), allocatable :: x
    allocate(character :: x)
    x = "1234567"
    print *, x
end program
```

**Comment by Thirumalai-Shaktivel** [comment]
Perfect! Now, only the string comparison issue remains..!

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 8 above)
Comment (certik) [reply src/lfortran/semantics/ast_body_visitor.cpp:929]: If you instrument this:
```fortran
program example
    implicit none
    character(:), allocatable :: x
    allocate(character :: x)
    print *, len(x)
    x = "1234567"
    print *, len(x)
    print *, x
end program
```
and run it:
```console
$ gfortran a.f90 && ./a.out 
           1
           7
 1234567
```
You can see that the line `allocate(character :: x)` allocates just size 1 string. Then the assignment `x = "1234567"` implicitly reallocates the LHS to 7.

So that's what we need to do as well in LFortran.

File: integration_tests/format_08.f90 [hunk 9]
```diff
@@ -6,4 +6,4 @@ program format_08
     b = 1
     print "(aai6)", a,"hi",15
     print "(1000(i6))", b
-end program
\ No newline at end of file
+end program
```
Comment (certik) [inline integration_tests/format_08.f90:9]: This is minor, but I would not do white space changes like that, send a separate PR. That way each PR is as small as possible, without distractions.

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 8 above)
Comment (Thirumalai-Shaktivel) [reply src/lfortran/semantics/ast_body_visitor.cpp:929]: Got it!

File: tests/reference/run-write5-4a89b6b.stdout [hunk 10]
```diff
@@ -1,3 +1,3 @@
 hibye
-first, last
+first     , last      
 Hello world!
```
Comment (certik) [inline tests/reference/run-write5-4a89b6b.stdout:3]: Note: This change is correct.

**Comment by certik** [comment]
This PR seems fine.

**Comment by Thirumalai-Shaktivel** [comment]
Prefect! This PR clears all the memory leaks related to strings in conv_lda for the 1st iteration.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 11]
```diff
@@ -201,6 +201,9 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     std::unique_ptr<LLVMSetInterface> set_api_sc;
     std::unique_ptr<LLVMArrUtils::Descriptor> arr_descr;
     std::vector<llvm::Value*> heap_arrays;
+    // std::map<array, size>
+    std::map<llvm::Value*, llvm::Value*> to_be_allocated;
+    Vec<llvm::Value*> to_be_deallocated;
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:206]: Is this only for strings? If so, let's name it `strings_to_be_allocated` and `strings_to_be_deallocated`.

**Review by certik — APPROVED** [review]
I think this is good.

I left one comment regarding renaming, otherwise fine. Thanks!

**Comment by certik** [comment]
Let's rebase on top of main, and clean up the history.

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
Thanks!
